### PR TITLE
[LLPC] Combine multiple const imm store instructions into single store

### PR DESCRIPTION
**Pull request by mariusz-sikora-at-amd**
Finds multiple store instructions which stores constant values into the same allocated memory. If all found stores are filling all allocated fields in the memory, then multiple stores are replaced with single store. This single store later may be promote to Global Variable.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit 3f4daea9e57c9e7a4a412a2917f23d33dd4e5d4e
<details><summary>CTS tests (Failed: 0/139021)</summary><ul><li>Built with version 1.3.5.2</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 41144/69517 (59.2%)</li>
<li>Failed: 0/69517 (0.0%)</li>
<li>Not Supported: 28373/69517 (40.8%)</li><li>Warnings: 0/69517 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 41178/69504 (59.2%)</li>
<li>Failed: 0/69504 (0.0%)</li>
<li>Not Supported: 28326/69504 (40.8%)</li><li>Warnings: 0/69504 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 8eeb7cc862db816b90cba0ff88b0e870c77debe6
<details><summary>CTS tests (Failed: 0/139021)</summary><ul><li>Built with version 1.3.5.2</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 41144/69517 (59.2%)</li>
<li>Failed: 0/69517 (0.0%)</li>
<li>Not Supported: 28373/69517 (40.8%)</li><li>Warnings: 0/69517 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 41178/69504 (59.2%)</li>
<li>Failed: 0/69504 (0.0%)</li>
<li>Not Supported: 28326/69504 (40.8%)</li><li>Warnings: 0/69504 (0.0%)</li>
</ul></details>


</ul></details>

File: llpc/lower/llpcSpirvLowerConstImmediateStore.cpp [hunk 1]
```diff
@@ -101,6 +103,133 @@ void SpirvLowerConstImmediateStore::processAllocaInsts(Function *func) {
...
+      storeList.push_back(storeInst);
+    } else if (auto *GEP = dyn_cast<GetElementPtrInst>(Inst)) {
+      if (GEP->hasAllConstantIndices() && GEP->getNumIndices() == 2) {
+        // Find more store instructions in GEP uses.
+        for (Use &GEPUse : Inst->uses()) {
+          Uses.push_back(&GEPUse);
+        }
+      }
+    } else if (!isa<LoadInst>(Inst) && !isAssumeLikeIntrinsic(Inst)) {
+      return;
```
Comment (tsymalla) [inline llpc/lower/llpcSpirvLowerConstImmediateStore.cpp:166]: Can you early opt-out when collecting the uses? If you stumble upon a Instruction that would branch here, then there is no need to iterate over all other users

File: llpc/lower/llpcSpirvLowerConstImmediateStore.cpp [hunk 2]
```diff
@@ -101,6 +103,133 @@ void SpirvLowerConstImmediateStore::processAllocaInsts(Function *func) {
...
+    return;
+
+  BasicBlock *parentBB = storeList.back()->getParent();
+  std::map<int, Constant *> idxConstMap;
+  // Iterate through all store instructions correlated with Alloca and check if
+  // we have enough data to create single store instruction which will initialize
+  // whole Alloca memory.
+  for (auto *storeInst : storeList) {
+    // For simplicity all stores have to be in the same Basic Block.
+    if (storeInst->getParent() != parentBB)
```
Comment (tsymalla) [inline llpc/lower/llpcSpirvLowerConstImmediateStore.cpp:181]: Can you early opt-out during collecting the stores if there is a store residing in another BB? You can just remember the last parent BB and return, if the current BB differs

File: llpc/lower/llpcSpirvLowerConstImmediateStore.cpp [hunk 3]
```diff
@@ -101,6 +103,133 @@ void SpirvLowerConstImmediateStore::processAllocaInsts(Function *func) {
...
+  Constant *newInitArray = ConstantArray::get(ArrayTy, NewInitVal);
+
+  IRBuilder<> Builder(storeList.back());
+  Builder.CreateAlignedStore(newInitArray, allocaInst, allocaInst->getAlign());
+
+  // Remove no longer needed store and GEP instructions.
+  while (!storeList.empty()) {
+    StoreInst *storeInst = storeList.pop_back_val();
+    GetElementPtrInst *GEP = dyn_cast<GetElementPtrInst>(storeInst->getPointerOperand());
+    bool hasOneUser = GEP ? GEP->hasOneUser() : false;
```
Comment (tsymalla) [inline llpc/lower/llpcSpirvLowerConstImmediateStore.cpp:225]: That variable is not necessary

File: llpc/test/shaderdb/general/TestCombineOfMultipleStoreInstructions.frag [hunk 4]
```diff
@@ -0,0 +1,87 @@
+#version 460
```
Comment (tsymalla) [inline llpc/test/shaderdb/general/TestCombineOfMultipleStoreInstructions.frag:1]: Can you add a test that shows how the IR changes?

**Comment by nhaehnle** [comment]
I took the liberty of cleaning up this PR in #2740, which I think should supersede this one.

**Comment by mariusz-sikora-at-amd** [comment]
> I took the liberty of cleaning up this PR in #2740, which I think should supersede this one.

Thanks @nhaehnle for the help ! Your changes look much better then my. I will close this PR.

File: llpc/test/shaderdb/general/TestCombineOfMultipleStoreInstructions.frag (hunk 4 above)
Comment (mariusz-sikora-at-amd) [reply llpc/test/shaderdb/general/TestCombineOfMultipleStoreInstructions.frag:1]: I will update the test when Nicolai will merge his PR #2740

**Pull request closed without merging**